### PR TITLE
Topic recognition

### PR DESCRIPTION
**Pull request by Terym2001**
Implementation of Improved UNet arcitecture for image segmentation of the ISIC 2018 dataset

student number: 48258791

My pull request provides a solution to recognition task 1.

The implemented arcitecture is a version of the arcitecture referenced in the recognition task. The modules have been changed from 3d to 2d to accomodate for training on the ISIC 2018 dataset. The softmax function has been changed to sigmoid as the recognition task asked for binary segmentation.

The result looks promising, as the network achieves an average score of 0.8. There should however still be conducted tweaking to find optimal training parameters. I think this would result in a better en result. Some segmentations still have a very low score. Tweaking of smooth parameters and adding wieghtd biases could help with this problem (further testing needed)

Files:
images/individual_dsc.png
images/model_arcitecture,png
images/segmentation_example_1.png
README.md
dataset.py
modules.py
predict.py
train.py
utils.py

dependencies:
pytorch (2.1.0)
pillow (10.0.1)
matplotlib (3.7.2)
numpy (1.24.3)

### CODE REVIEW

**Comment by Terym2001** [comment]
UPDATE: the average dice score of 0.8 was achieved on the trainset and not the testset.

**Comment by wangzhaomxy** [comment]
This is an initial inspection, no action is required at this point.

- Well organized files.
- Algorithm solves the problem appropriately. Average Dice Coefficient in testing dataset: 0.8. Predicted Mask output looks good.
- Correctly using PyTorch to construct the Improved U-Net models and functions.
- Good coding design. 
- Sufficient docstring and comment in the code, but need some references.
- Easy difficulty problem, no higher difficulty marks.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Easy -10
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable, missing usage -1
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
